File: backend/app/user_ui_preferences.py
"""Normalize and merge per-user UI preferences (layout, filters, column widths)."""

from __future__ import annotations

import uuid
from typing import Any, Literal

from pydantic import BaseModel, Field

CalendarView = Literal["dayGridMonth", "timeGridWeek", "timeGridDay", "listWeek"]
TaskLayout = Literal["list", "kanban"]
TaskSortKey = Literal["reference", "client", "matter", "task", "date", "assigned", "priority"]
MainMenuSortKey = Literal["reference", "client", "matter", "feeEarner", "status", "created"]
ContactsSortKey = Literal["name", "type", "email", "phone"]
CaseStatusFilter = Literal["", "open", "closed", "archived", "quote", "post_completion"]
SortDir = Literal["asc", "desc"]

_CALENDAR_VIEWS = frozenset({"dayGridMonth", "timeGridWeek", "timeGridDay", "listWeek"})
_TASK_LAYOUTS = frozenset({"list", "kanban"})
_TASK_SORT_KEYS = frozenset({"reference", "client", "matter", "task", "date", "assigned", "priority"})
_MAIN_MENU_SORT_KEYS = frozenset({"reference", "client", "matter", "feeEarner", "status", "created"})
_CONTACTS_SORT_KEYS = frozenset({"name", "type", "email", "phone"})
_CASE_STATUS_FILTERS = frozenset({"", "open", "closed", "archived", "quote", "post_completion"})
_SORT_DIRS = frozenset({"asc", "desc"})

MAIN_MENU_COLUMN_WIDTHS_DEFAULT = [110, 240, 300, 180, 100]
TASKS_MENU_COLUMN_WIDTHS_DEFAULT = [90, 66, 90, 210, 300, 183, 73]
CONTACTS_COLUMN_WIDTHS_DEFAULT = [270, 210, 210, 210]

_LEGACY_AUTO_MAIN_MENU_COLUMN_WIDTHS = MAIN_MENU_COLUMN_WIDTHS_DEFAULT
_LEGACY_AUTO_TASKS_MENU_COLUMN_WIDTHS = TASKS_MENU_COLUMN_WIDTHS_DEFAULT
_LEGACY_AUTO_CONTACTS_COLUMN_WIDTHS = CONTACTS_COLUMN_WIDTHS_DEFAULT

_SEARCH_MAX_LEN = 500


class UserUiPreferencesOut(BaseModel):
    calendar_view: CalendarView = "dayGridMonth"
    case_calendar_view: CalendarView = "dayGridMonth"
    tasks_menu_layout: TaskLayout = "list"
    case_tasks_layout: TaskLayout = "list"
    tasks_menu_sort_key: TaskSortKey = "priority"
    tasks_menu_sort_dir: SortDir = "asc"
    case_tasks_sort_key: TaskSortKey = "priority"
    case_tasks_sort_dir: SortDir = "asc"
    main_menu_sort_key: MainMenuSortKey = "created"
    main_menu_sort_dir: SortDir = "desc"
    main_menu_search: str = ""
    main_menu_filter_matter_type: str = ""
    main_menu_filter_fee_earner_user_id: str = ""
    main_menu_filter_case_status: CaseStatusFilter = ""
    main_menu_filter_matter_types: list[str] = Field(default_factory=list)
    main_menu_filter_fee_earner_user_ids: list[str] = Field(default_factory=list)
    main_menu_filter_case_statuses: list[CaseStatusFilter] = Field(default_factory=list)
    tasks_menu_search: str = ""
    tasks_menu_filter_matter_type: str = ""
    contacts_search: str = ""
    contacts_sort_key: ContactsSortKey = "name"
    contacts_sort_dir: SortDir = "asc"
    calendar_selected_calendar_ids: list[str] = Field(default_factory=list)
    main_menu_column_widths: list[int] = Field(default_factory=list)
    tasks_menu_column_widths: list[int] = Field(default_factory=list)
    contacts_column_widths: list[int] = Field(default_factory=list)


class UserUiPreferencesPatch(BaseModel):
    calendar_view: CalendarView | None = None
    case_calendar_view: CalendarView | None = None
    tasks_menu_layout: TaskLayout | None = None
    case_tasks_layout: TaskLayout | None = None
    tasks_menu_sort_key: TaskSortKey | None = None
    tasks_menu_sort_dir: SortDir | None = None
    case_tasks_sort_key: TaskSortKey | None = None
    case_tasks_sort_dir: SortDir | None = None
    main_menu_sort_key: MainMenuSortKey | None = None
    main_menu_sort_dir: SortDir | None = None
    main_menu_search: str | None = None
    main_menu_filter_matter_type: str | None = None
    main_menu_filter_fee_earner_user_id: str | None = None
    main_menu_filter_case_status: CaseStatusFilter | None = None
    main_menu_filter_matter_types: list[str] | None = None
    main_menu_filter_fee_earner_user_ids: list[str] | None = None
    main_menu_filter_case_statuses: list[CaseStatusFilter] | None = None
    tasks_menu_search: str | None = None
    tasks_menu_filter_matter_type: str | None = None
    contacts_search: str | None = None
    contacts_sort_key: ContactsSortKey | None = None
    contacts_sort_dir: SortDir | None = None
    calendar_selected_calendar_ids: list[str] | None = None
    main_menu_column_widths: list[int] | None = None
    tasks_menu_column_widths: list[int] | None = None
    contacts_column_widths: list[int] | None = None


def _pick_str(raw: Any, allowed: frozenset[str], default: str) -> str:
    if isinstance(raw, str) and raw in allowed:
        return raw
    return default


def _pick_search(raw: Any) -> str:
    if not isinstance(raw, str):
        return ""
    return raw.strip()[:_SEARCH_MAX_LEN]


def _pick_filter_text(raw: Any) -> str:
    if not isinstance(raw, str):
        return ""
    return raw.strip()[:200]


def _pick_fee_earner_id(raw: Any) -> str:
    if not isinstance(raw, str):
        return ""
    s = raw.strip()
    if not s:
        return ""
    try:
        uuid.UUID(s)
    except ValueError:
        return ""
    return s


def _normalize_id_list(raw: Any) -> list[str]:
    if not isinstance(raw, list):
        return []
    out: list[str] = []
    for item in raw:
        if isinstance(item, str) and item.strip():
            out.append(item.strip())
    return out


def _clamp_widths(raw: Any, expected: int) -> list[int] | None:
    if not isinstance(raw, list) or len(raw) != expected:
        return None
    out: list[int] = []
    for item in raw:
        if not isinstance(item, (int, float)):
            return None
        try:
            w = int(item)
        except (TypeError, ValueError):
            return None
        out.append(max(48, min(2000, w)))
    return out


def _stored_column_widths(raw: Any, expected: int, legacy_auto: list[int]) -> list[int]:
    clamped = _clamp_widths(raw, expected)
    if clamped is None:
        return []
    if clamped == legacy_auto:
        return []
    return clamped


def _pick_filter_text_list(raw: Any) -> list[str]:
    if not isinstance(raw, list):
        return []
    out: list[str] = []
    for item in raw:
        s = _pick_filter_text(item)
        if s and s not in out:
            out.append(s)
    return out


def _pick_fee_earner_id_list(raw: Any) -> list[str]:
    if not isinstance(raw, list):
        return []
    out: list[str] = []
    for item in raw:
        s = _pick_fee_earner_id(item)
        if s and s not in out:
            out.append(s)
    return out


def _pick_case_status_list(raw: Any) -> list[str]:
    if not isinstance(raw, list):
        return []
    out: list[str] = []
    for item in raw:
        if isinstance(item, str) and item in _CASE_STATUS_FILTERS and item != "":
            if item not in out:
                out.append(item)
    return out


def _legacy_main_menu_matter_types(data: dict[str, Any]) -> list[str]:
    from_list = _pick_filter_text_list(data.get("main_menu_filter_matter_types"))
    if from_list:
        return from_list
    one = _pick_filter_text(data.get("main_menu_filter_matter_type"))
    return [one] if one else []


def _legacy_main_menu_fee_earner_ids(data: dict[str, Any]) -> list[str]:
    from_list = _pick_fee_earner_id_list(data.get("main_menu_filter_fee_earner_user_ids"))
    if from_list:
        return from_list
    one = _pick_fee_earner_id(data.get("main_menu_filter_fee_earner_user_id"))
    return [one] if one else []


def _legacy_main_menu_case_statuses(data: dict[str, Any]) -> list[str]:
    from_list = _pick_case_status_list(data.get("main_menu_filter_case_statuses"))
    if from_list:
        return from_list
    one = _pick_str(data.get("main_menu_filter_case_status"), _CASE_STATUS_FILTERS, "")
    return [one] if one else []


def user_ui_preferences_out(stored: dict[str, Any] | None) -> UserUiPreferencesOut:
    data = stored if isinstance(stored, dict) else {}
    return UserUiPreferencesOut(
        calendar_view=_pick_str(data.get("calendar_view"), _CALENDAR_VIEWS, "dayGridMonth"),  # type: ignore[arg-type]
        case_calendar_view=_pick_str(data.get("case_calendar_view"), _CALENDAR_VIEWS, "dayGridMonth"),  # type: ignore[arg-type]
        tasks_menu_layout=_pick_str(data.get("tasks_menu_layout"), _TASK_LAYOUTS, "list"),  # type: ignore[arg-type]
        case_tasks_layout=_pick_str(data.get("case_tasks_layout"), _TASK_LAYOUTS, "list"),  # type: ignore[arg-type]
        tasks_menu_sort_key=_pick_str(data.get("tasks_menu_sort_key"), _TASK_SORT_KEYS, "priority"),  # type: ignore[arg-type]
        tasks_menu_sort_dir=_pick_str(data.get("tasks_menu_sort_dir"), _SORT_DIRS, "asc"),  # type: ignore[arg-type]
        case_tasks_sort_key=_pick_str(data.get("case_tasks_sort_key"), _TASK_SORT_KEYS, "priority"),  # type: ignore[arg-type]
        case_tasks_sort_dir=_pick_str(data.get("case_tasks_sort_dir"), _SORT_DIRS, "asc"),  # type: ignore[arg-type]
        main_menu_sort_key=_pick_str(data.get("main_menu_sort_key"), _MAIN_MENU_SORT_KEYS, "created"),  # type: ignore[arg-type]
        main_menu_sort_dir=_pick_str(data.get("main_menu_sort_dir"), _SORT_DIRS, "desc"),  # type: ignore[arg-type]
        main_menu_search=_pick_search(data.get("main_menu_search")),
        main_menu_filter_matter_type=_pick_filter_text(data.get("main_menu_filter_matter_type")),
        main_menu_filter_fee_earner_user_id=_pick_fee_earner_id(data.get("main_menu_filter_fee_earner_user_id")),
        main_menu_filter_case_status=_pick_str(data.get("main_menu_filter_case_status"), _CASE_STATUS_FILTERS, ""),  # type: ignore[arg-type]
        main_menu_filter_matter_types=_legacy_main_menu_matter_types(data),
        main_menu_filter_fee_earner_user_ids=_legacy_main_menu_fee_earner_ids(data),
        main_menu_filter_case_statuses=_legacy_main_menu_case_statuses(data),  # type: ignore[arg-type]
        tasks_menu_search=_pick_search(data.get("tasks_menu_search")),
        tasks_menu_filter_matter_type=_pick_filter_text(data.get("tasks_menu_filter_matter_type")),
        contacts_search=_pick_search(data.get("contacts_search")),
        contacts_sort_key=_pick_str(data.get("contacts_sort_key"), _CONTACTS_SORT_KEYS, "name"),  # type: ignore[arg-type]
        contacts_sort_dir=_pick_str(data.get("contacts_sort_dir"), _SORT_DIRS, "asc"),  # type: ignore[arg-type]
        calendar_selected_calendar_ids=_normalize_id_list(data.get("calendar_selected_calendar_ids")),
        main_menu_column_widths=_stored_column_widths(
            data.get("main_menu_column_widths"),
            len(MAIN_MENU_COLUMN_WIDTHS_DEFAULT),
            _LEGACY_AUTO_MAIN_MENU_COLUMN_WIDTHS,
        ),
        tasks_menu_column_widths=_stored_column_widths(
            data.get("tasks_menu_column_widths"),
            len(TASKS_MENU_COLUMN_WIDTHS_DEFAULT),
            _LEGACY_AUTO_TASKS_MENU_COLUMN_WIDTHS,
        ),
        contacts_column_widths=_stored_column_widths(
            data.get("contacts_column_widths"),
            len(CONTACTS_COLUMN_WIDTHS_DEFAULT),
            _LEGACY_AUTO_CONTACTS_COLUMN_WIDTHS,
        ),
    )


def merge_ui_preferences_patch(
    stored: dict[str, Any] | None,
    patch: UserUiPreferencesPatch,
) -> dict[str, Any]:
    merged = user_ui_preferences_out(stored).model_dump()
    updates = patch.model_dump(exclude_unset=True)
    if "main_menu_search" in updates and updates["main_menu_search"] is not None:
        updates["main_menu_search"] = _pick_search(updates["main_menu_search"])
    if "tasks_menu_search" in updates and updates["tasks_menu_search"] is not None:
        updates["tasks_menu_search"] = _pick_search(updates["tasks_menu_search"])
    if "contacts_search" in updates and updates["contacts_search"] is not None:
        updates["contacts_search"] = _pick_search(updates["contacts_search"])
    if "main_menu_filter_matter_type" in updates and updates["main_menu_filter_matter_type"] is not None:
        updates["main_menu_filter_matter_type"] = _pick_filter_text(updates["main_menu_filter_matter_type"])
    if "tasks_menu_filter_matter_type" in updates and updates["tasks_menu_filter_matter_type"] is not None:
        updates["tasks_menu_filter_matter_type"] = _pick_filter_text(updates["tasks_menu_filter_matter_type"])
    if "main_menu_filter_fee_earner_user_id" in updates and updates["main_menu_filter_fee_earner_user_id"] is not None:
        updates["main_menu_filter_fee_earner_user_id"] = _pick_fee_earner_id(
            updates["main_menu_filter_fee_earner_user_id"]
        )
    if "main_menu_filter_matter_types" in updates and updates["main_menu_filter_matter_types"] is not None:
        updates["main_menu_filter_matter_types"] = _pick_filter_text_list(updates["main_menu_filter_matter_types"])
    if "main_menu_filter_fee_earner_user_ids" in updates and updates["main_menu_filter_fee_earner_user_ids"] is not None:
        updates["main_menu_filter_fee_earner_user_ids"] = _pick_fee_earner_id_list(
            updates["main_menu_filter_fee_earner_user_ids"]
        )
    if "main_menu_filter_case_statuses" in updates and updates["main_menu_filter_case_statuses"] is not None:
        updates["main_menu_filter_case_statuses"] = _pick_case_status_list(updates["main_menu_filter_case_statuses"])
    if "calendar_selected_calendar_ids" in updates and updates["calendar_selected_calendar_ids"] is not None:
        updates["calendar_selected_calendar_ids"] = _normalize_id_list(updates["calendar_selected_calendar_ids"])
    if "main_menu_column_widths" in updates and updates["main_menu_column_widths"] is not None:
        clamped = _clamp_widths(updates["main_menu_column_widths"], len(MAIN_MENU_COLUMN_WIDTHS_DEFAULT))
        if clamped is not None:
            updates["main_menu_column_widths"] = clamped
        else:
            updates.pop("main_menu_column_widths", None)
    if "tasks_menu_column_widths" in updates and updates["tasks_menu_column_widths"] is not None:
        clamped = _clamp_widths(updates["tasks_menu_column_widths"], len(TASKS_MENU_COLUMN_WIDTHS_DEFAULT))
        if clamped is not None:
            updates["tasks_menu_column_widths"] = clamped
        else:
            updates.pop("tasks_menu_column_widths", None)
    if "contacts_column_widths" in updates and updates["contacts_column_widths"] is not None:
        clamped = _clamp_widths(updates["contacts_column_widths"], len(CONTACTS_COLUMN_WIDTHS_DEFAULT))
        if clamped is not None:
            updates["contacts_column_widths"] = clamped
        else:
            updates.pop("contacts_column_widths", None)
    merged.update(updates)
    return user_ui_preferences_out(merged).model_dump()


def ui_preferences_is_default(prefs: UserUiPreferencesOut) -> bool:
    return prefs == UserUiPreferencesOut()

File: frontend/src/userUiPreferences.ts
/** Per-user UI layout preferences synced with the server. */

export type CalendarView = 'dayGridMonth' | 'timeGridWeek' | 'timeGridDay' | 'listYear'
export type TaskLayout = 'list' | 'kanban'
export type TaskSortKey = 'reference' | 'client' | 'matter' | 'task' | 'date' | 'assigned' | 'priority'
export type MainMenuSortKey = 'reference' | 'client' | 'matter' | 'feeEarner' | 'status' | 'source' | 'created'
export type ContactsSortKey = 'name' | 'type' | 'email' | 'phone'
export type CaseStatusFilter = '' | 'open' | 'closed' | 'archived' | 'quote' | 'quote_closed' | 'post_completion'
export type MainMenuCaseStatusFilter = Exclude<CaseStatusFilter, ''>
export type SortDir = 'asc' | 'desc'

export const MAIN_MENU_COLUMN_WIDTHS_DEFAULT = [110, 165, 300, 130, 225] as const
export const TASKS_MENU_COLUMN_WIDTHS_DEFAULT = [90, 66, 90, 210, 300, 183, 73] as const
export const CONTACTS_COLUMN_WIDTHS_DEFAULT = [270, 210, 210, 210] as const

export const MAIN_MENU_COLUMN_COUNT = MAIN_MENU_COLUMN_WIDTHS_DEFAULT.length
export const TASKS_MENU_COLUMN_COUNT = TASKS_MENU_COLUMN_WIDTHS_DEFAULT.length
export const CONTACTS_COLUMN_COUNT = CONTACTS_COLUMN_WIDTHS_DEFAULT.length

export type UserUiPreferences = {
  calendar_view: CalendarView
  case_calendar_view: CalendarView
  tasks_menu_layout: TaskLayout
  case_tasks_layout: TaskLayout
  tasks_menu_sort_key: TaskSortKey
  tasks_menu_sort_dir: SortDir
  case_tasks_sort_key: TaskSortKey
  case_tasks_sort_dir: SortDir
  main_menu_sort_key: MainMenuSortKey
  main_menu_sort_dir: SortDir
  main_menu_search: string
  main_menu_filter_matter_type: string
  main_menu_filter_fee_earner_user_id: string
  main_menu_filter_case_status: CaseStatusFilter
  main_menu_filter_matter_types: string[]
  main_menu_filter_fee_earner_user_ids: string[]
  main_menu_filter_case_statuses: MainMenuCaseStatusFilter[]
  tasks_menu_search: string
  tasks_menu_filter_matter_type: string
  contacts_search: string
  contacts_sort_key: ContactsSortKey
  contacts_sort_dir: SortDir
  calendar_selected_calendar_ids: string[]
  main_menu_column_widths: number[]
  tasks_menu_column_widths: number[]
  contacts_column_widths: number[]
}

export const DEFAULT_UI_PREFERENCES: UserUiPreferences = {
  calendar_view: 'dayGridMonth',
  case_calendar_view: 'dayGridMonth',
  tasks_menu_layout: 'list',
  case_tasks_layout: 'list',
  tasks_menu_sort_key: 'priority',
  tasks_menu_sort_dir: 'asc',
  case_tasks_sort_key: 'priority',
  case_tasks_sort_dir: 'asc',
  main_menu_sort_key: 'created',
  main_menu_sort_dir: 'desc',
  main_menu_search: '',
  main_menu_filter_matter_type: '',
  main_menu_filter_fee_earner_user_id: '',
  main_menu_filter_case_status: '',
  main_menu_filter_matter_types: [],
  main_menu_filter_fee_earner_user_ids: [],
  main_menu_filter_case_statuses: [],
  tasks_menu_search: '',
  tasks_menu_filter_matter_type: '',
  contacts_search: '',
  contacts_sort_key: 'name',
  contacts_sort_dir: 'asc',
  calendar_selected_calendar_ids: [],
  main_menu_column_widths: [],
  tasks_menu_column_widths: [],
  contacts_column_widths: [],
}

const CACHE_KEY = 'canary.uiPreferences.v2'

/** Dispatched after menu column widths are reset so all tables revert immediately. */
export const MENU_COLUMN_RESET_EVENT = 'canary-menu-columns-reset'

export function notifyMenuColumnReset(): void {
  window.dispatchEvent(new Event(MENU_COLUMN_RESET_EVENT))
}
export const LEGACY_TASKS_MENU_LAYOUT_KEY = 'canary.tasks.menuLayout'
const SEARCH_MAX = 500

const CALENDAR_VIEWS = new Set<CalendarView>(['dayGridMonth', 'timeGridWeek', 'timeGridDay', 'listYear'])

function pickCalendarView(raw: unknown, fallback: CalendarView): CalendarView {
  if (typeof raw === 'string' && raw === 'listWeek') return 'listYear'
  return pick(raw, CALENDAR_VIEWS, fallback)
}
const TASK_LAYOUTS = new Set<TaskLayout>(['list', 'kanban'])
const TASK_SORT_KEYS = new Set<TaskSortKey>([
  'reference',
  'client',
  'matter',
  'task',
  'date',
  'assigned',
  'priority',
])
const MAIN_MENU_SORT_KEYS = new Set<MainMenuSortKey>([
  'reference',
  'client',
  'matter',
  'feeEarner',
  'status',
  'source',
  'created',
])
const CONTACTS_SORT_KEYS = new Set<ContactsSortKey>(['name', 'type', 'email', 'phone'])
const CASE_STATUS_FILTERS = new Set<CaseStatusFilter>([
  '',
  'open',
  'closed',
  'archived',
  'quote',
  'quote_closed',
  'post_completion',
])
const SORT_DIRS = new Set<SortDir>(['asc', 'desc'])
const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i

function pick<T extends string>(raw: unknown, allowed: Set<T>, fallback: T): T {
  return typeof raw === 'string' && allowed.has(raw as T) ? (raw as T) : fallback
}

function pickSearch(raw: unknown): string {
  return typeof raw === 'string' ? raw.trim().slice(0, SEARCH_MAX) : ''
}

function pickFilterText(raw: unknown): string {
  return typeof raw === 'string' ? raw.trim().slice(0, 200) : ''
}

function pickFeeEarnerId(raw: unknown): string {
  if (typeof raw !== 'string') return ''
  const s = raw.trim()
  return UUID_RE.test(s) ? s : ''
}

function pickFilterTextList(raw: unknown): string[] {
  if (!Array.isArray(raw)) return []
  const out: string[] = []
  for (const item of raw) {
    const s = pickFilterText(item)
    if (s && !out.includes(s)) out.push(s)
  }
  return out
}

function pickFeeEarnerIdList(raw: unknown): string[] {
  if (!Array.isArray(raw)) return []
  const out: string[] = []
  for (const item of raw) {
    const s = pickFeeEarnerId(item)
    if (s && !out.includes(s)) out.push(s)
  }
  return out
}

function pickCaseStatusList(raw: unknown): MainMenuCaseStatusFilter[] {
  if (!Array.isArray(raw)) return []
  const out: MainMenuCaseStatusFilter[] = []
  for (const item of raw) {
    if (typeof item === 'string' && item !== '' && CASE_STATUS_FILTERS.has(item as CaseStatusFilter)) {
      const v = item as MainMenuCaseStatusFilter
      if (!out.includes(v)) out.push(v)
    }
  }
  return out
}

function legacyMainMenuMatterTypes(data: Record<string, unknown>): string[] {
  const fromList = pickFilterTextList(data.main_menu_filter_matter_types)
  if (fromList.length) return fromList
  const one = pickFilterText(data.main_menu_filter_matter_type)
  return one ? [one] : []
}

function legacyMainMenuFeeEarnerIds(data: Record<string, unknown>): string[] {
  const fromList = pickFeeEarnerIdList(data.main_menu_filter_fee_earner_user_ids)
  if (fromList.length) return fromList
  const one = pickFeeEarnerId(data.main_menu_filter_fee_earner_user_id)
  return one ? [one] : []
}

function legacyMainMenuCaseStatuses(data: Record<string, unknown>): MainMenuCaseStatusFilter[] {
  const fromList = pickCaseStatusList(data.main_menu_filter_case_statuses)
  if (fromList.length) return fromList
  const one = pick(
    data.main_menu_filter_case_status,
    CASE_STATUS_FILTERS,
    DEFAULT_UI_PREFERENCES.main_menu_filter_case_status,
  )
  return one ? [one as MainMenuCaseStatusFilter] : []
}

function normalizeIdList(raw: unknown): string[] {
  if (!Array.isArray(raw)) return []
  return raw.filter((x): x is string => typeof x === 'string' && x.trim().length > 0).map((x) => x.trim())
}

function normalizeWidths(raw: unknown, expected: number): number[] {
  if (!Array.isArray(raw) || raw.length !== expected) return []
  const out: number[] = []
  for (const item of raw) {
    if (typeof item !== 'number' || !Number.isFinite(item)) return []
    out.push(Math.max(48, Math.min(2000, Math.round(item))))
  }
  return out
}

export function normalizeUiPreferences(raw: unknown): UserUiPreferences {
  const data = raw && typeof raw === 'object' ? (raw as Record<string, unknown>) : {}
  return {
    calendar_view: pickCalendarView(data.calendar_view, DEFAULT_UI_PREFERENCES.calendar_view),
    case_calendar_view: pickCalendarView(data.case_calendar_view, DEFAULT_UI_PREFERENCES.case_calendar_view),
    tasks_menu_layout: pick(data.tasks_menu_layout, TASK_LAYOUTS, DEFAULT_UI_PREFERENCES.tasks_menu_layout),
    case_tasks_layout: pick(data.case_tasks_layout, TASK_LAYOUTS, DEFAULT_UI_PREFERENCES.case_tasks_layout),
    tasks_menu_sort_key: pick(data.tasks_menu_sort_key, TASK_SORT_KEYS, DEFAULT_UI_PREFERENCES.tasks_menu_sort_key),
    tasks_menu_sort_dir: pick(data.tasks_menu_sort_dir, SORT_DIRS, DEFAULT_UI_PREFERENCES.tasks_menu_sort_dir),
    case_tasks_sort_key: pick(data.case_tasks_sort_key, TASK_SORT_KEYS, DEFAULT_UI_PREFERENCES.case_tasks_sort_key),
    case_tasks_sort_dir: pick(data.case_tasks_sort_dir, SORT_DIRS, DEFAULT_UI_PREFERENCES.case_tasks_sort_dir),
    main_menu_sort_key: pick(data.main_menu_sort_key, MAIN_MENU_SORT_KEYS, DEFAULT_UI_PREFERENCES.main_menu_sort_key),
    main_menu_sort_dir: pick(data.main_menu_sort_dir, SORT_DIRS, DEFAULT_UI_PREFERENCES.main_menu_sort_dir),
    main_menu_search: pickSearch(data.main_menu_search),
    main_menu_filter_matter_type: pickFilterText(data.main_menu_filter_matter_type),
    main_menu_filter_fee_earner_user_id: pickFeeEarnerId(data.main_menu_filter_fee_earner_user_id),
    main_menu_filter_case_status: pick(
      data.main_menu_filter_case_status,
      CASE_STATUS_FILTERS,
      DEFAULT_UI_PREFERENCES.main_menu_filter_case_status,
    ),
    main_menu_filter_matter_types: legacyMainMenuMatterTypes(data),
    main_menu_filter_fee_earner_user_ids: legacyMainMenuFeeEarnerIds(data),
    main_menu_filter_case_statuses: legacyMainMenuCaseStatuses(data),
    tasks_menu_search: pickSearch(data.tasks_menu_search),
    tasks_menu_filter_matter_type: pickFilterText(data.tasks_menu_filter_matter_type),
    contacts_search: pickSearch(data.contacts_search),
    contacts_sort_key: pick(data.contacts_sort_key, CONTACTS_SORT_KEYS, DEFAULT_UI_PREFERENCES.contacts_sort_key),
    contacts_sort_dir: pick(data.contacts_sort_dir, SORT_DIRS, DEFAULT_UI_PREFERENCES.contacts_sort_dir),
    calendar_selected_calendar_ids: normalizeIdList(data.calendar_selected_calendar_ids),
    main_menu_column_widths: normalizeWidths(
      data.main_menu_column_widths,
      MAIN_MENU_COLUMN_COUNT,
    ),
    tasks_menu_column_widths: normalizeWidths(
      data.tasks_menu_column_widths,
      TASKS_MENU_COLUMN_COUNT,
    ),
    contacts_column_widths: normalizeWidths(
      data.contacts_column_widths,
      CONTACTS_COLUMN_COUNT,
    ),
  }
}

export function uiPreferencesEqual(a: UserUiPreferences, b: UserUiPreferences): boolean {
  return JSON.stringify(a) === JSON.stringify(b)
}

export function readCachedUiPreferences(): UserUiPreferences {
  try {
    const raw = localStorage.getItem(CACHE_KEY)
    if (raw) return normalizeUiPreferences(JSON.parse(raw))
  } catch {
    /* ignore */
  }
  return { ...DEFAULT_UI_PREFERENCES }
}

export function writeCachedUiPreferences(prefs: UserUiPreferences): void {
  try {
    localStorage.setItem(CACHE_KEY, JSON.stringify(prefs))
  } catch {
    /* ignore */
  }
}

export function legacyUiPreferenceOverrides(): Partial<UserUiPreferences> {
  const out: Partial<UserUiPreferences> = {}
  try {
    const legacyTasks = localStorage.getItem(LEGACY_TASKS_MENU_LAYOUT_KEY)
    if (legacyTasks === 'kanban') out.tasks_menu_layout = 'kanban'
  } catch {
    /* ignore */
  }
  return out
}

export async function persistUserUiPreferences(
  token: string,
  patch: Partial<UserUiPreferences>,
): Promise<UserUiPreferences> {
  const { apiFetch } = await import('./api')
  const user = await apiFetch<{ ui_preferences?: UserUiPreferences }>('/users/me/ui-preferences', {
    token,
    method: 'PUT',
    json: patch,
  })
  const prefs = normalizeUiPreferences(user.ui_preferences)
  writeCachedUiPreferences(prefs)
  return prefs
}

/** Clear saved column widths for all resizable menu tables (main menu, quotes, tasks, contacts). */
export async function resetMenuColumnWidths(token: string): Promise<UserUiPreferences> {
  const prefs = await persistUserUiPreferences(token, {
    main_menu_column_widths: [],
    tasks_menu_column_widths: [],
    contacts_column_widths: [],
  })
  notifyMenuColumnReset()
  return prefs
}
